### PR TITLE
ARTEMIS-1327 - Support checked exceptions in ActiveMQServerPlugin

### PR DESCRIPTION
**Pull request by cshannon**
This will allow plugin writers to use checked exceptions when writing
plugins

### CODE REVIEW

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/server/impl/RemotingServiceImpl.java [hunk 1]
```diff
@@ -514,8 +514,11 @@ public void connectionCreated(final ActiveMQComponent component,
       }
 
       ConnectionEntry entry = protocol.createConnectionEntry((Acceptor) component, connection);
-      server.callBrokerPlugins(server.hasBrokerPlugins() ? plugin -> plugin.afterCreateConnection(entry.connection) : null);
-
+      try {
+         server.callBrokerPlugins(server.hasBrokerPlugins() ? plugin -> plugin.afterCreateConnection(entry.connection) : null);
+      } catch (Exception e) {
+         throw new IllegalStateException("Error executing afterCreateConnection plugin method: " + e.getMessage(), e);
```
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/server/impl/RemotingServiceImpl.java]: shouldn't you just log on this case...

I would expect any failures on the broker plugin to just be logged.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/server/impl/RemotingServiceImpl.java [hunk 2]
```diff
@@ -537,7 +540,11 @@ public void connectionDestroyed(final Object connectionID) {
       if (conn != null && !conn.connection.isSupportReconnect()) {
          RemotingConnection removedConnection = removeConnection(connectionID);
          if (removedConnection != null) {
-            server.callBrokerPlugins(server.hasBrokerPlugins() ? plugin -> plugin.afterDestroyConnection(removedConnection) : null);
+            try {
+               server.callBrokerPlugins(server.hasBrokerPlugins() ? plugin -> plugin.afterDestroyConnection(removedConnection) : null);
+            } catch (Exception e) {
+               throw new IllegalStateException("Error executing afterDestroyConnection plugin method: " + e.getMessage(), e);
```
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/server/impl/RemotingServiceImpl.java]: same here.. I would just log.warn... you cannot interrupt the broker if someone deploys a bugged interceptor.. can you?

Comment (cshannon) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/server/impl/RemotingServiceImpl.java]: Yep, good point, i'll switch it and push up a new version.

**Comment by cshannon** [comment]
I pushed up a new version that logs the error a the warn level

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/server/impl/RemotingServiceImpl.java (hunk 2 above)
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/server/impl/RemotingServiceImpl.java]: make it catch Throwable :)

Comment (cshannon) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/server/impl/RemotingServiceImpl.java]: done

**Comment by cshannon** [comment]
@clebertsuconic - So I just realized that I don't think we want to swallow exceptions.  If there's an exception in the plugin don't we want it to propagate?  There are 2 cases for exceptions, the first is a buggy plugin and that just means the plugin should be fixed.  The second is if there is an actual error and I think that error should be propagated and not just logged.

For example, one of my use cases for this is I want to validate clientIds.  So I want to throw an exception and reject a clientId in the beforeSessionMetadataAdded method if it doesn't meet some criteria and have it propagated back to the client.  (There's actually another Jira I need to open to make this work properly to make sure the exception is thrown back to the client).

**Comment by cshannon** [comment]
@clebertsuconic - Or another use case...say you want to write a custom authentication plugin.  You would normally want to throw something like a security exception, etc in the plugin methods but now you can't do that.

**Comment by clebertsuconic** [comment]
What about expose activemqexception.  But swallow others?


Let me amend my PR and will and you feedback.

**Comment by cshannon** [comment]
That would probably be ok, just as long there is some way to propagate an exception when the plugin writer wants to.  If we do that should we change the plugin interface to throw ActiveMQException instead of Exception?  I'm about to head out for the day but I'll take a look at your new PR in the morning.

**Comment by clebertsuconic** [comment]
@cshannon take a look on #1446 please?

**Comment by clebertsuconic** [comment]
@cshannon I will merge this now as I need to add something on the server plugins... 

if there's anything wrong with it.. we can review it later, ok?

**Comment by cshannon** [comment]
@clebertsuconic - This looks good to me

**Pull request closed without merging**